### PR TITLE
Walking Hammerdin bugfix

### PR DESCRIPTION
**Pull request by drallcom3**
The Hammerdin often accidentally clicked on monsters while walking to the end location. He did so while having Vigor active (because the bot thinks he's still walking). The existing fix to activate Concentration early did not work. Now it does.
You might have noticed many complaints about it in Discord.

If you want to test:
Do Eldritch or Trav 10x without the fix as a Hammerdin without any form of teleport. You will notice him casting hammers with Vigor active at the beginning.
With my fix that doesn't happen anymore.

Eld, Shenk, Trav, Pindle fixed. The others not, because they're impossible without teleport anyway.

### CODE REVIEW

**Comment by D2RLegit** [comment]
Not at a pc but it looks like you took out vigor altogether for the run and replaced it with concentration? Have you verified this still works with a hdin wearing a enigma?

**Comment by drallcom3** [comment]
Vigor is still being use like before, up until the safe spot right before a boss. Just the last few meters are done with Concentration.
The bot was supposed to do that all along, but the functionality was broken.

Works fine with teleport charges.
I don't have an Enigma, but I don't see why it won't work. I haven't touched that part. Someone has to test it of course.

**Comment by drallcom3** [comment]
I realized that I could just test it in singleplayer, so I did that. No issues with Enigma. Teleports happily around.

File: src/char/hammerdin.py [hunk 1]
```diff
@@ -51,12 +51,6 @@ def pre_buff(self):
         mouse.click(button="right")
         wait(self._cast_duration, self._cast_duration + 0.06)
 
-    def on_capabilities_discovered(self, capabilities: CharacterCapabilities):
-        # In case we have a running pala, we want to switch to concentration when moving to the boss
-        # ass most likely we will click on some mobs and already cast hammers
-        if capabilities.can_teleport_natively:
-            self._do_pre_move = False
-
```
Comment (mgleed) [inline src/char/hammerdin.py:59]: Why was this function removed?

Comment (D2RLegit) [reply src/char/hammerdin.py:59]: I was wondering the same which is why I asked if it still works for hdin with tele. Ive been running as a walkadin and seems fine. Ill double verify in a bit if enigma still works.

**Comment by D2RLegit** [comment]
[0.7.3-dev 2022-05-09 06:47:41,781] INFO       Capabilities: CharacterCapabilities(can_teleport_natively=False, can_teleport_with_charges=True)

Populates when using walkadin for pindle.
Walkadin does not work for pindle in this pr.

File: src/char/hammerdin.py (hunk 1 above)
Comment (drallcom3) [reply src/char/hammerdin.py:59]: When debugging self._do_pre_move it turned out that it reverts back to true. The Hammerdin class logic expected it to stay false though. It was the main culprit of why the Hammerdin seemed to "attack" while having Vigor turned on. It should have switched to Concentration, but didn't. Not relying on self._do_pre_move fixed it immediately and then the on_capabilities_discovered wasn't required anymore.

Also it lead to unneccesarily complicated logic constructs that were hard to follow and could be done in a simpler and more readable way.

**Comment by drallcom3** [comment]
I get:
INFO       Capabilities: CharacterCapabilities(can_teleport_natively=False, can_teleport_with_charges=False)
and when adding checks further down the line:
DEBUG      capabilities.can_teleport_natively: False
DEBUG      capabilities.can_teleport_with_charges: False
when doing Pindle without any teleport equipped.

Did you get that message while having a teleport charge item equipped and ready?

I can only achieve
INFO       Capabilities: CharacterCapabilities(can_teleport_natively=False, can_teleport_with_charges=True)
if I have a teleport item with charges equipped and a hotkey assigned in-game and a hotkey assigned in the config. Which is (I assume) cirrect behaviour. If you did the test with a teleport charge item equipped, you're not a Walkadin anymore. I just ran the test. With teleport charges the Hammerdin behaves exactly like in master. Which is correct behaviour, since this PR is about walking, not teleport charges. The Hammerdin needs some more fixes, as you have discovered, but I didn't want to make this PR too massive.

In short: I can't see a bug here (that didn't already exist).

**Comment by D2RLegit** [comment]
It saying i have teleport charges when i run pindle even though im not using any teleport or teleport charge items. it should be a walkdin and its being recognized that it has teleport charges.

**Comment by drallcom3** [comment]
Thanks.
If your self.capabilities.can_teleport_with_charges were incorrectly true, that would be a bug well beyond the Hammerdin file (as it's not being set in there and the info shows up before any boss functions are executed).

**Comment by D2RLegit** [comment]
Pretty sure its because you removed 

    def on_capabilities_discovered(self, capabilities: CharacterCapabilities):
        # In case we have a running pala, we want to switch to concentration when moving to the boss
        # ass most likely we will click on some mobs and already cast hammers
        if capabilities.can_teleport_natively:
            self._do_pre_move = False

but i could be wrong. My walkdin works fine on every other pr.

**Comment by drallcom3** [comment]
What exactly does happen to your Hammerdin when you do Pindle with this PR?

**Comment by drallcom3** [comment]
It's actually this #758 bug.
The Hammerdin just made us discover it.

File: src/char/hammerdin.py (hunk 1 above)
Comment (VladimirMakaev) [reply src/char/hammerdin.py:59]: I wouldn't touch this code. It can easily break something for Enigma / Tele-Charge paladin . And you need to test 3 different version of paladin on all routes to make sure it's sound:
1) Walkdadin Trav/Shenk/...
2) Enigma Paladin Trav/Shenk/...
3) Tele-Charge Paladin Trav/Shenk

Comment (VladimirMakaev) [reply src/char/hammerdin.py:59]: You should just cast Concentration and move with PreMove=False that way you won't change to Vigor. 
preMove() will flip to Vigor in case of Walkadin or Tele-Charge (there is no way to pass that telecharge is to be used based on current code structure). So in case you're opting to tele-charge then you should manage PreMove part manually. E.g. casting concentration and doing PreMove=False, TeleCharge = True or PreMove=True if you want Vigor to be on for walking / running out of tele charges and walking.

**Review by VladimirMakaev — CHANGES_REQUESTED** [review]

File: src/char/hammerdin.py (hunk 1 above)
Comment (drallcom3) [reply src/char/hammerdin.py:59]: > I wouldn't touch this code. It can easily break something for Enigma / Tele-Charge paladin . And you need to test 3 different version of paladin on all routes to make sure it's sound:
> 
>     1. Walkdadin Trav/Shenk/...
> 
>     2. Enigma Paladin Trav/Shenk/...
> 
>     3. Tele-Charge Paladin Trav/Shenk

It didn't even work the way it was intended to work. No need to re-introduce bugs.

This PR has been tested extensively with all three variants. Some people even happily and successfully use it for their daily runs.

Comment (drallcom3) [reply src/char/hammerdin.py:59]: > You should just cast Concentration and move with PreMove=False that way you won't change to Vigor. preMove() will flip to Vigor in case of Walkadin or Tele-Charge (there is no way to pass that telecharge is to be used based on current code structure). So in case you're opting to tele-charge then you should manage PreMove part manually. E.g. casting concentration and doing PreMove=False, TeleCharge = True or PreMove=True if you want Vigor to be on for walking / running out of tele charges and walking.

Which is what I did.
